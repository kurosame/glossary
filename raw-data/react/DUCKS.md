## category

react

## titles

Ducks
Redux

## description

Redux のデザインパターン  
Action や Reducer を modules にパッケージする

vuejs-boilerplate でも使ってる気がする  
1 ファイルにはしてないけど、actions, getters, mutations, state を import して modules でパッケージして export してる  
そして Store 作るときに modules をセットしている
