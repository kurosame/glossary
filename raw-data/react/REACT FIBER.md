## category

react

## titles

React Fiber
react-fiber

## description

Begin, Complete, Commit のフェーズがある

### Begin, Complete

Component 単位で実行される  
componentWillMount, componentWillReceiveProps, shouldComponentUpdate, componentWillUpdate  
非同期レンダリングでは中断・再開により、複数回ライフサイクル関数が呼ばれる可能性がある

### Commit

全ての Fiber が完了した時に実行される  
DOM 更新などの副作用が発生する処理はここでまとめて反映される  
componentDidMount, componentDidUpdate, componentWillUnmount

### Time Slicing

Fiber により画面のレンダリングが非同期になると、Fiber 単位で処理の優先度を付けることができる  
例えばユーザ操作の優先度を高くすることで、画面がレンダリング中でもユーザ操作を優先して処理することが可能となる

### Suspense

非同期通信をしている間、画面のレンダリングを中断できる  
非同期通信が完了した後に、画面をレンダリングすることが可能  
render に Promise を投げて、その Promise が完了した後に render するみたいな？
