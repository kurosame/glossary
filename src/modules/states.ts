import type { LoginState } from '@/modules/login'
import type { WordState } from '@/modules/word'

export interface States {
  login: LoginState
  words: WordState[]
}
