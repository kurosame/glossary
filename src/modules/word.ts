import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const GET_WORDS = 'GET_WORDS'
export const SET_WORDS = 'SET_WORDS'

export interface IWordActions {
  getWords: ActionFunctionAny<Action<{}>>
  setWords: ActionFunctionAny<Action<{ words: IWordState[] }>>
}

export interface IWordState {
  id: string
  category: string
  titles: string[]
  description: string
}

const initialState: IWordState[] = []

export const { getWords, setWords } = createActions(GET_WORDS, SET_WORDS)

export const words: Reducer<IWordState[], Action<IWordState[]>> = handleActions(
  {
    [SET_WORDS]: (
      state: IWordState[],
      action: Action<IWordState[]>
    ): IWordState[] => {
      return !!action.payload.length ? action.payload : state
    }
  },
  initialState
)
