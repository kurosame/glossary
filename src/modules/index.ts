import { ICounterActions, ICounterState } from '@/modules/counter'
import { IWordActions, IWordState } from '@/modules/word'

export interface IActions {
  counter: ICounterActions
  word: IWordActions
}

export interface IStates {
  counter: ICounterState
  words: IWordState[]
}
