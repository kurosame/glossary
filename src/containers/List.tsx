import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Dispatch } from 'redux'
import Words from '@/components/Words'
import { States } from '@/modules/states'
import { WordState, GET_WORDS } from '@/modules/word'

interface Props {
  match: { params: { category: string } }
}

const useGetWords = (ws: WordState[]): void => {
  const dispatch = useDispatch<Dispatch>()
  useEffect(() => {
    if (!ws.length) dispatch({ type: GET_WORDS })
  }, [ws, dispatch])
}

const List: React.FC<Props> = p => {
  const isLogin = useSelector<States, boolean>(s => s.login.isLogin)
  const words = useSelector<States, WordState[]>(s => s.words)

  useGetWords(words)

  if (!isLogin) return null
  return (
    <Words
      words={words.filter(
        w => !p.match.params.category || w.category === p.match.params.category
      )}
    />
  )
}

export default List
