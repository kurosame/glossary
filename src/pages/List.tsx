import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators, Dispatch } from 'redux'
import Words from '@/components/Words'
import { States } from '@/modules/states'
import { LoginState } from '@/modules/login'
import { getWords, WordActions, WordState } from '@/modules/word'

interface Props {
  state: { login: LoginState; words: WordState[] }
  actions: Pick<WordActions, 'getWords'>
  match: { params: { category: string } }
  store?: unknown
}

export class List extends React.PureComponent<Props> {
  constructor(props: Props) {
    super(props)
    if (!props.state.words.length) {
      props.actions.getWords()
    }
  }

  public render(): JSX.Element | null {
    if (!this.props.state.login.isLogin) return null
    return (
      <Words
        words={this.props.state.words.filter(
          w =>
            !this.props.match.params.category ||
            w.category === this.props.match.params.category
        )}
      />
    )
  }
}

export default connect(
  (states: States) => ({ state: { login: states.login, words: states.words } }),
  (dispatch: Dispatch) => ({
    actions: {
      getWords: bindActionCreators(getWords, dispatch)
    }
  })
)(List)
