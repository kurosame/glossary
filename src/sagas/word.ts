import { collection, getDocs } from 'firebase/firestore'
import type { Action } from 'redux-actions'
import type { CallEffect, PutEffect, TakeEffect } from 'redux-saga/effects'
import { call, put, take } from 'redux-saga/effects'

import { firestore } from '@/firebase/index'
import type { WordState } from '@/modules/word'
import { GET_WORDS, setWords } from '@/modules/word'

export const getFirestoreWords = (): Promise<WordState[]> =>
  getDocs(
    collection(firestore(), 'words').withConverter<WordState>({
      fromFirestore: (s, o) => ({
        id: s.data(o).id,
        category: s.data(o).category,
        titles: s.data(o).titles,
        description: s.data(o).description
      }),
      toFirestore: w => ({ id: w.id, category: w.category, titles: w.titles, description: w.description })
    })
  )
    .then(res =>
      res.docs.map(d => ({
        id: d.id,
        category: d.data().category,
        titles: d.data().titles,
        description: d.data().description
      }))
    )
    .catch((err: Error) => {
      console.error(`GET_WORDS Firestore response error: ${err.message}`)
      return []
    })

export function* getWords(): IterableIterator<TakeEffect | CallEffect | PutEffect<Action<{ words: WordState[] }>>> {
  while (true) {
    yield take(GET_WORDS)
    yield put(setWords!({ words: yield call(getFirestoreWords) }))
  }
}
