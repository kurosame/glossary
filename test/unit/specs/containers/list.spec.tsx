import React from 'react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import List from '@/containers/List'
import { WordState, GET_WORDS } from '@/modules/word'
import { States } from '@/modules/states'
import { render, cleanup, RenderResult } from '@testing-library/react'
import '@testing-library/jest-dom'

let mockDispatch: jest.Mock
let wrapper: (
  isLogin: boolean,
  words: WordState[],
  match: { params: { category: string } }
) => RenderResult
beforeEach(() => {
  mockDispatch = jest.fn()
  wrapper = (
    isLogin: boolean,
    words: WordState[],
    match: { params: { category: string } }
  ): RenderResult => {
    const store = configureStore<States>()({
      login: { isLogin },
      words
    })
    store.dispatch = mockDispatch
    return render(
      <Provider store={store}>
        <List match={match} />
      </Provider>
    )
  }
})
afterEach(() => {
  cleanup()
  jest.restoreAllMocks()
})

test('Call `getWords` when `words` is empty', () => {
  wrapper(true, [], { params: { category: '' } })

  expect(mockDispatch).toBeCalled()
  expect(mockDispatch.mock.calls[0][0]).toEqual({ type: GET_WORDS })
  expect(mockDispatch.mock.calls[1]).toBeUndefined()
})

test('Not Call `getWords` when `words` exists', () => {
  wrapper(
    true,
    [
      {
        id: 'React',
        category: 'react',
        titles: ['React'],
        description: 'It a React'
      }
    ],
    { params: { category: '' } }
  )

  expect(mockDispatch).not.toBeCalled()
})

test('Render DOM if `isLogin` is true', () => {
  expect(
    wrapper(true, [], { params: { category: '' } }).container.querySelector(
      'div'
    )
  ).not.toBeNull()
})

test('Not render DOM if `isLogin` is false', () => {
  expect(
    wrapper(false, [], { params: { category: '' } }).container.querySelector(
      'div'
    )
  ).toBeNull()
})

test('Output all `words` when `category` is empty', () => {
  expect(
    wrapper(
      true,
      [
        {
          id: 'React',
          category: 'react',
          titles: ['React'],
          description: 'It a React'
        },
        {
          id: 'Vue.js',
          category: 'vue',
          titles: ['Vue', 'Vuex'],
          description: 'It a Vue'
        }
      ],
      { params: { category: '' } }
    ).container
  ).toHaveTextContent('ReactReactIt a ReactVue.jsVueVuexIt a Vue')
})

test('Output empty when `category` exists but not match', () => {
  expect(
    wrapper(
      true,
      [
        {
          id: 'React',
          category: 'react',
          titles: ['React'],
          description: 'It a React'
        },
        {
          id: 'Vue.js',
          category: 'vue',
          titles: ['Vue', 'Vuex'],
          description: 'It a Vue'
        }
      ],
      { params: { category: 'angular' } }
    ).container
  ).toHaveTextContent('')
})

test('Match the snapshot', () => {
  expect(
    wrapper(
      true,
      [
        {
          id: 'React',
          category: 'react',
          titles: ['React'],
          description: 'It a React'
        },
        {
          id: 'Vue.js',
          category: 'vue',
          titles: ['Vue', 'Vuex'],
          description: 'It a Vue'
        }
      ],
      { params: { category: '' } }
    ).asFragment()
  ).toMatchSnapshot()
})
