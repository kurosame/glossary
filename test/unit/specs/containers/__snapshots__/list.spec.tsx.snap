// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot when category exists 1`] = `"<ul class=\\"MuiList-root-1 MuiList-padding-2\\"><li class=\\"MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12\\"><div class=\\"sc-bwzfXH jwfMIB\\"><div class=\\"MuiPaper-root-16 MuiPaper-elevation1-19 MuiPaper-rounded-17 MuiCard-root-15\\"><div class=\\"MuiCardHeader-root-43\\" data-test=\\"id\\"><div class=\\"MuiCardHeader-content-46\\"><span class=\\"MuiTypography-root-49 MuiTypography-headline-54 MuiCardHeader-title-47\\"><b></b></span></div></div><div class=\\"MuiCardContent-root-73\\" data-test=\\"titles\\"></div><div class=\\"MuiCardContent-root-73\\" data-test=\\"description\\"><div></div></div></div></div></li></ul>"`;

exports[`Match the snapshot when category is nothing 1`] = `"<ul class=\\"MuiList-root-1 MuiList-padding-2\\"><li class=\\"MuiListItem-root-5 MuiListItem-default-8 MuiListItem-gutters-12\\"><div class=\\"sc-bwzfXH jwfMIB\\"><div class=\\"MuiPaper-root-16 MuiPaper-elevation1-19 MuiPaper-rounded-17 MuiCard-root-15\\"><div class=\\"MuiCardHeader-root-43\\" data-test=\\"id\\"><div class=\\"MuiCardHeader-content-46\\"><span class=\\"MuiTypography-root-49 MuiTypography-headline-54 MuiCardHeader-title-47\\"><b></b></span></div></div><div class=\\"MuiCardContent-root-73\\" data-test=\\"titles\\"></div><div class=\\"MuiCardContent-root-73\\" data-test=\\"description\\"><div></div></div></div></div></li></ul>"`;
